src/mac.m

```diff
@@ -1174,23 +1174,21 @@ - (void)windowDidResignKey:(NSNotification*)notification
 puglSendEvent(PuglView* view, const PuglEvent* event)
 {
   if (event->type == PUGL_CLIENT) {
-    PuglWrapperView* wrapper = view->impl->wrapperView;
-    const NSWindow*  window  = [wrapper window];
-    const NSRect     rect    = [wrapper frame];
-    const NSPoint    center  = {NSMidX(rect), NSMidY(rect)};
-
-    NSEvent* nsevent =
-      [NSEvent otherEventWithType:NSApplicationDefined
-                         location:center
-                    modifierFlags:0
-                        timestamp:[[NSProcessInfo processInfo] systemUptime]
-                     windowNumber:window.windowNumber
-                          context:nil
-                          subtype:PUGL_CLIENT
-                            data1:(NSInteger)event->client.data1
-                            data2:(NSInteger)event->client.data2];
-
-    [view->world->impl->app postEvent:nsevent atStart:false];
+    PuglEvent copiedEvent = *event;
+
+    CFRunLoopObserverRef observer = CFRunLoopObserverCreateWithHandler(
+      NULL,
+      kCFRunLoopAfterWaiting,
+      false,
+      0,
+      ^(CFRunLoopObserverRef, CFRunLoopActivity) {
+        puglDispatchEvent(view, &copiedEvent);
+      });
+
+    CFRunLoopAddObserver(CFRunLoopGetMain(), observer, kCFRunLoopCommonModes);
```

Comment:
According to the documentation, this does nothing if there is already an observer registered for the same mode.  Doesn't this mean multiple client events won't work?  Does it make sense to add a run loop observer for every event anyway?

Reply:
Are you referring to this part of the docs?
> If rl already contains observer in mode, this function does nothing.

I interpreted that to mean that the function was a no-op if _that specific observer_ had already been added in the specified mode. Perhaps I can test this by loading a couple of DPF plugins in REAPER and checking whether client events get sent for both.

Reply:
I've tested this out a bit in some modified DPF plugins. I added two calls to puglSendEvent inside puglUpdate (which I knew was guaranteed to be called regularly by the host), both with different client events. Then, I stuck a breakpoint inside the block passed to `CFRunLoopObserverCreateWithHandler`. In LLDB, I can see both events getting triggered, which indicates that adding the second observer is not a no-op, and does not remove the first observer.

I also tried loading up multiple plugin instances with the same modification, and checked that both received the client event callbacks. It appears that all plugin instances which register observers for a particular mode receive the callbacks that they register.

Reply:
Cool, thanks for looking into it.

The "timing" of this approach seems weird to me, since more or less the whole point of `puglUpdate` is to process all events _now_, but I'll get into my Mac at some point this weekend and tinker with it a bit.  It might not matter since MacOS is already wildly divergent from the other platforms WRT event timing/order semantics anyway...